src/clj/rems/logging.clj

```diff
@@ -0,0 +1,58 @@
+(ns rems.logging
+  (:require [clojure.test :refer [deftest is testing]])
+  (:import (org.slf4j MDC)))
+
+(defn with-mdc* [m f]
+  (let [m (->> m
+               (map (fn [[key val]]
+                      [(if (keyword? key)
+                         (name key)
+                         key)
+                       (str val)]))
+               (into {}))]
+    (try
+      (doseq [[key val] m]
+        (when-not (empty? val)
+          (MDC/put key val)))
+      (f)
+      (finally
+        (doseq [key (keys m)]
+          (MDC/remove key))))))
+
+(defmacro with-mdc
+  "Adds a map of values to SLF4J's MDC for logging context."
+  [m & body]
+  `(with-mdc* ~m (fn [] ~@body)))
+
+(deftest test-with-mdc
+  (testing "adds keys to context"
+    (with-mdc {"foo" "bar"}
+      (is (= "bar" (MDC/get "foo")))))
+  (testing "removes keys from context afterwards"
+    (with-mdc {"foo" "bar"})
+    (is (nil? (MDC/get "foo"))))
+  (testing "can add multiple keys at once"
+    (with-mdc {"foo" "one"
+               "bar" "two"}
+      (is (= "one" (MDC/get "foo")))
+      (is (= "two" (MDC/get "bar"))))
+    (is (empty? (MDC/getCopyOfContextMap))))
+  (testing "converts keyword keys to string"
+    (with-mdc {:foo "bar"}
+      (is (= "bar" (MDC/get "foo")))))
+  (testing "converts non-string values to string"
```

Comment:
Here you could also use `are` for the multiple tests.

Reply:
Using a template expression for just two asserts seems very overkill. It took me over a minute to read and understand that what `are` does.

Reply:
There are at least three here alone and I'm sure you can think of even more if you look at it.

Reply:
Honestly this test looks like word porridge to me now.

Reply:
In this project there are 825 usages of `is` and only 1 usage of `are` (in `rems.test.db.applications/test-handling-event?`). That's a rather obscure operation. Complicating these simple asserts with that would reduce the test readability considerably.

Reply:
`are` is a standard Clojure test structure and not at all obscure. This is something to what I would write

```
(deftest test-with-mdc
  (with-mdc {"foo" "one"
             "bar" "two"}
    (is (= "one" (MDC/get "foo")))
    (is (= "two" (MDC/get "bar"))))

  (is (empty? (MDC/getCopyOfContextMap)) "empty outside sexp")
  (is (nil? (MDC/get "foo")) "key does not exist outside sexp")

  (testing "converts values to string"
    (are [expected x] (= expected (with-mdc {"foo" x} (MDC/get "foo")))
      ":bar" :bar
      "123" 123
      "true" true))

  (with-mdc {"foo" nil} (is (empty? (MDC/getCopyOfContextMap)) "nil should be ignored"))
  (with-mdc {"foo" ""} (is (empty? (MDC/getCopyOfContextMap)) "empty string should be ignored"))
  (with-mdc {"foo" false} (is (= "false" (MDC/get "foo")) "false should not be ignored")))
```

Reply:
Here's another variant

```
(deftest test-with-mdc
  (with-mdc {"foo" "one"
             "bar" "two"}
    (is (= "one" (MDC/get "foo")))
    (is (= "two" (MDC/get "bar")))
    (is (nil? (MDC/get "foobar")) "non-existant key does not exist"))

  (is (nil? (MDC/get "foo")) "key does not exist outside sexp")

  (are [expected x] (= expected (with-mdc {"foo" x} (MDC/get "foo")))
    ":bar" :bar
    "123" 123
    "false" false
    nil nil
    nil ""))
```

Reply:
Does one still get with one what row the fail was, eg. if there are 5 conditions in are-clause, does one get accurate information which one of them failed? If yes, then I think are is ok.

Reply:
For the string conversion case `are` makes sense. I'll refactor that.

Reply:
Actually, when there are multiple failures, `are` reports them all but it does not tell that which of the test cases had the problem. In the following example all the assertion failures refer to line 54, which is the line where `are` starts, but the test cases are on lines 56 and 57. 

Especially when there are many test cases, it becomes harder to know what caused the failure. The failure message shows only the expected and actual values, but not the input value, so when there are multiple input values which produce the same output, it's not possible to know which of them failed unless you comment out the other test cases and try them one by one.
```clj
FAIL in (test-with-mdc) (logging.clj:54)
converts non-string values to string
expected: ":bar1"
  actual: ":bar"
    diff: - ":bar1"
          + ":bar"

FAIL in (test-with-mdc) (logging.clj:54)
converts non-string values to string
expected: "123x"
  actual: "123"
    diff: - "123x"
          + "123"
```

Reply:
That was my worry, so it would seem more practical to use if?

Reply:
Yes, seems like it.